oci/private/container.bzl

```diff
@@ -3,102 +3,79 @@ _attrs = {
     "base": attr.string(
         mandatory = True,
     ),
-
     # See: https://github.com/opencontainers/image-spec/blob/main/config.md#properties
     "entrypoint": attr.string_list(),
     "cmd": attr.string_list(),
     "labels": attr.string_list(),
     "tag": attr.string_list(),
     "layers": attr.label_list(),
+    "_container_sh_tpl": attr.label(default = "container.sh.tpl", allow_single_file = True),
 }
 
-def _strip_external(path):
-    return path[len("external/"):] if path.startswith("external/") else path
-
 def _impl(ctx):
-    toolchain = ctx.toolchains["@contrib_rules_oci//oci:crane_toolchain_type"]
-
-    launcher = ctx.actions.declare_file("crane.sh")
+    crane = ctx.toolchains["@contrib_rules_oci//oci:crane_toolchain_type"]
+    registry = ctx.toolchains["@contrib_rules_oci//oci:registry_toolchain_type"]
 
-    # TODO: dynamically get --platform from toolchain
-    ctx.actions.write(
-        launcher,
-        """#!/usr/bin/env bash
-set -euo pipefail
-{crane} $@""".format(
-            crane = toolchain.crane_info.crane_path,
-        ),
+    launcher = ctx.actions.declare_file("container_%s.sh" % ctx.label.name)
+    ctx.actions.expand_template(
+        template = ctx.file._container_sh_tpl,
+        output = launcher,
         is_executable = True,
+        substitutions = {
+            "{registry_launcher_path}": registry.registry_info.launcher_path,
+            "{crane_path}": crane.crane_info.crane_path,
+            "{storage_dir}": "/".join([ctx.bin_dir.path, ctx.label.package, "storage_%s" % ctx.label.name])
+        }
     )
 
-    # Pull the image
-    pull = ctx.actions.args()
-
-    tar = ctx.actions.declare_file("base_%s.tar" % ctx.label.name)
+    inputs = []
 
-    pull.add_all([
-        "append",
-        "--base",
+    args = ctx.actions.args()
+    args.add_all([
+        "mutate",
         ctx.attr.base,
-        "--output",
-        tar,
-        "--new_tag",
-        ctx.label.name,
+        "--tag",
+        "oci:registry/%s" % ctx.label.name
     ])
 
-    inputs = list()
-
-    inputs.extend(toolchain.crane_info.crane_files)
-
     if ctx.attr.layers:
-        pull.add("--new_layer")
+        args.add("--append")
         for layer in ctx.attr.layers:
+            # TODO: allow only .tar files
             inputs.extend(layer[DefaultInfo].files.to_list())
```

Comment:
This is not necessary, see my next comment. The issue is a performance impact on converting to a list.